### ISSUE TITLE
Lock-free bag 

### ISSUE THREAD

**Issue by bartoszmodelski** [ocaml-multicore/lockfree#29]
[Bag (or multiset)](https://en.wikipedia.org/wiki/Multiset) is a data structure, which stores a collection of values. Elements are inserted and removed one at a time. In contrast with queue or stack, bag has no ordering. That is _remove_ can return the youngest, oldest or any other item currently in the bag. In principle, lack of ordering eliminates contention points and should lead to better throughput and scalability.

It would be useful to have one here: 
* Bag would provide better throughput than stack (`ws_deque`) for workloads without significant locality between tasks. 
* Better performance for Reagents (https://github.com/ocaml-multicore/reagents/issues/4)

I'm not sure what's the best design. [A lock-free algorithm for concurrent bags](https://dl.acm.org/doi/10.1145/1989493.1989550) proposes a compelling one.

**Comment by Johan511** [comment]
Hi, I am interested in working on this issue and have been trying to implement the threadBlocks as described in the paper. One issue I have faced was with the Backoff module, any advice on how I can import it into my Bag.ml file.

Also I would like some advice on how to allot a threadBlock to each thread.

**Comment by polytypic** [comment]
I totally agree that a scalable lock-free bag would be a useful data structure to have.

> Also I would like some advice on how to allot a threadBlock to each thread.

You probably want to use [DLS or Domain Local Storage](https://github.com/ocaml/ocaml/blob/af72270e1a7597c8b3f1158308f0001546459095/stdlib/domain.mli#L91-L143).

> Better throughput for Domainslib than with stack or queue (for workloads without significant locality between tasks).

I'm curious to see the result.  For performing tasks, however, I realised many years ago that stack like or LIFO ordering tends to be generally preferable.  The reason isn't contention.  The reason is that LIFO ordering results in a [Depth-First Search](https://en.wikipedia.org/wiki/Depth-first_search) like behaviour, while FIFO results in a [Breadth-First Search](https://en.wikipedia.org/wiki/Breadth-first_search) like behaviour.  The crucial difference is in memory and general resource usage.  Deviations from the DFS behaviour tend to increase resource usage.

In a kind of build tool that performed large numbers of largely independent tasks in parallel I initially used a queue.  However, that resulted in the program running out of memory.  Memory usage grew proportional to the number of tasks that could proceed if given CPU time.  I changed the program to use a stack.  Memory usage then stayed roughly (constant or) proportional to the number of CPU cores.

**Comment by lyrm** [comment]
We now have a bag implemented in `Saturn`, which uses a hash table internally.

However, I’m keeping this issue open, as it would be interesting to explore how a different algorithm or implementation might compare in terms of performance.

Since this is less of a priority compared to other new data structure issues, I am changing the tag from `new data structure` to `enhancement`.
